### PR TITLE
Fixes part of #59: Build app module binding adapters + layouts with Bazel 

### PR DESCRIPTION
**Pull request by miaboloix**
## Explanation
Fixes part of #59: Build app module binding adapters with Bazel [BLOCKED: #1581]

In order to build app module in Bazel, parts of the module must be built before others in order to avoid circular dependencies and in order for data-binding to work. In this PR, I am building the module's binding adapters and the module's layout files. 

See the graph below:
Screen Shot 2020-08-18 at 8 48 49 PM

This is a visual representation of the libraries app module will have and how they will depend on each other. The libraries in green have already been implemented. [See #1581](https://github.com/oppia/oppia-android/pull/1581). The libraries in red are implemented in this PR and the libraries in grey are implemented in a PR to follow this one.

Here is a quick summary of each library's function:
- **listeners**: Builds any listener files needed for views or view models
- **annotations**: Builds any scope annotation or helper files needed for the views or view models
- **resources**: Builds non-layout resource files  
- **views**: Builds view files
- **view models**: Builds view model files
- **databinding_adapters** and **recyclerview_adapters**: Builds the custom binding adapter files. Note that these two libraries have since been combined into **binding_adapters**.
- **databinding_resources**: Builds layout files
- **app**: The complete app module library

**NOTES:** 
1. Gradle does not build with this implementation yet but there are plans to write a custom macro to fix this issue.
2. If you have not set up Bazel for data-binding in Oppia Android, please follow the instructions laid out in [this document.](https://docs.google.com/document/d/1V0RMehfi3GLVsZ6dyin1nTo4_TO28LACp4wPlEnvnM0/edit?usp=sharing)

To build `databinding_adapters`, run 
```
~/bazel-databinding-new/bazel-bin/src/bazel build //:databinding_adapters --android_databinding_use_v3_4_args --experimental_android_databinding_v2 --override_repository=android_tools=/tmp/android_tools --java_header_compilation=false
```
To build `recyclerview_adapters`, run 
```
~/bazel-databinding-new/bazel-bin/src/bazel build //:recyclerview_adapters --android_databinding_use_v3_4_args --experimental_android_databinding_v2 --override_repository=android_tools=/tmp/android_tools --java_header_compilation=false
```
To build `databinding_resources`, run 
```
~/bazel-databinding-new/bazel-bin/src/bazel build //:databinding_resources --android_databinding_use_v3_4_args --experimental_android_databinding_v2 --override_repository=android_tools=/tmp/android_tools --java_header_compilation=false
```
 

## Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [ ] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [ ] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [ ] The PR is made from a branch that's **not** called "develop".
- [ ] The PR is made from a branch that is up-to-date with "develop".
- [ ] The PR's branch is based on "develop" and not on any other branch.
- [ ] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: Does ``"src/main/res/layout*/**"`` work to replace all four?

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: These different names are a bit confusing (``binding`` vs ``databinding``)--please rename or consolidate to clarify.

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: Here & elsewhere: does any package in app actually need to be public other than the top-level library included in the ``android_binary`` target?

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: The inconsistent custom packages is a bit confusing--can we use a consistent pattern for these?

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel:498]: Please keep the ordering consistent, specifically I believe local libraries are listed before top-level ones, and those should be before external workspace imports which are before artifacts.

Ditto elsewhere.

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: Please keep deps lists sorted lexicographically, as well.

Ditto elsewhere.

File: app/src/main/DbAdaptersAndroidManifest.xml
Comment (BenHenning) [inline app/src/main/DbAdaptersAndroidManifest.xml]: File name: prefer spelling out the thing rather than abbreviating. Ditto for the next manifest.

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: Similar to my comment in your other PR, but can we reuse the altered list below to avoid duplicating the adapters in two places?

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: This documentation is a bit confusing--we're altering the imports to account for the altered package & file name? Can this be made a bit clearer?

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @miaboloix! Overall, no major issues with this PR. Just a few comments.

File: app/BUILD.bazel
Comment (vinitamurthi) [inline app/BUILD.bazel]: Can we set the visibility per BUILD file (i.e. package defaults) instead of per target? Also, do these need to be public? Can we instead specify exactly which BUILD files need these as dependencies?

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: Yes, just replaced it!

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: I altered quite a few names - let me know if you think it helps!

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: Yep!

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: Fixed!

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: Fixed! They can all be private except for `app`. :)

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel:498]: Fixed!

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: Fixed!

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: These do not need to be public - the only library that should be public is `app`. However, if you look at the bottom of the file, where the firebase dependencies are, the package default visibility is stated there. Because some of the firebase rules do not have visibility attributes (at least that you can set) and they need to be public (since other libraries depend on them) -- the defualt visibility of this file needs to be public.

File: app/src/main/DbAdaptersAndroidManifest.xml
Comment (miaboloix) [reply app/src/main/DbAdaptersAndroidManifest.xml]: Done!

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: Fixed! I basically have the package name mirroring the library name - but I just replaced `_` with `.`  :)

**Comment by BenHenning** [comment]
Holding off on reviewing until the branch is fixed. Please reassign once that's done @miaboloix.

File: app/BUILD.bazel
Comment (vinitamurthi) [reply app/BUILD.bazel]: Okay got it , thanks!

**Review by vinitamurthi — APPROVED** [review]

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: Perhaps: ``DATABINDING_LAYOUTS``?

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: Please remove merge artifact. Ditto elsewhere in the PR.

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: prefer ``enable_data_binding = True`` here & everywhere for consistency & clarity.

File: app/src/main/java/org/oppia/app/player/audio/LanguageInterface.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/audio/LanguageInterface.kt]: Unlike the shim interfaces, this one is longer-lived. Let's describe what implementations should actually do for this method, e.g.: ``Returns whether the user is actively seeking to a new audio position, that is, dragging the knob to a new position in the audio track.``

Ditto for the method below.

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughActivityListener.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/walkthrough/WalkthroughActivityListener.kt]: Methods should be verbs or verb phrases, e.g. ``moveToPreviousPage``. Also, both interfaces & their methods should have KDocs per my earlier comment about making it clear how implementations should implement the new functionality. Be sure to also explain circumstances, and document how callers should decide what parameter values to pass in.

Ditto for other new/updated interfaces in this PR.

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughPageChanger.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/walkthrough/WalkthroughPageChanger.kt]: Maybe: ``... should change pages during the app walkthrough flow.`` This isn't so generic to represent all multi-page activities, just the walkthrough. We should make that clear in the documentation.

Ditto for other interfaces. Make sure to be specific about what sort of effect the interface should have on the app.

File: app/src/main/java/org/oppia/app/onboarding/OnboardingListener.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/onboarding/OnboardingListener.kt]: Hmm, 'OnboardingListener' sounds like it will indicate whenever onboarding state has changed, but that's not actually what's happening here. Can we name this more specifically? Maybe 'OnboardingNavigationListener'?

File: app/BUILD.bazel
Comment (BenHenning) [reply app/BUILD.bazel]: This is much clearer--thanks!

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @miaboloix! Most of my comments are regarding clarity & documentation.

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: Done

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: Done

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: Done

File: app/src/main/java/org/oppia/app/onboarding/OnboardingListener.kt
Comment (miaboloix) [reply app/src/main/java/org/oppia/app/onboarding/OnboardingListener.kt]: Okay!

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughPageChanger.kt
Comment (miaboloix) [reply app/src/main/java/org/oppia/app/walkthrough/WalkthroughPageChanger.kt]: Okay!

File: app/src/main/java/org/oppia/app/player/audio/LanguageInterface.kt
Comment (miaboloix) [reply app/src/main/java/org/oppia/app/player/audio/LanguageInterface.kt]: Ah - will do. Thank you!

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughActivityListener.kt
Comment (miaboloix) [reply app/src/main/java/org/oppia/app/walkthrough/WalkthroughActivityListener.kt]: Okay!

File: app/src/main/java/org/oppia/app/player/audio/LanguageInterface.kt
Comment (miaboloix) [reply app/src/main/java/org/oppia/app/player/audio/LanguageInterface.kt]: Okay!

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughActivityListener.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/walkthrough/WalkthroughActivityListener.kt]:
```suggestion
   * @param currentProgress Integer describing the current page a user is on.
```

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughPageChanger.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/walkthrough/WalkthroughPageChanger.kt]:
```suggestion
/**
 * Listener for when an activity should change pages during the app walkthrough flow's welcome
 * fragment.
 */
```

Notice the formatting--we always format KDocs like this unless they can fit on one line. See: https://developer.android.com/kotlin/style-guide#formatting_2.

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: Here & elsewhere: let's make sure it's clearly documented what each list is meant to contain. Consider the perspective of members of the team: if they are adding a new class, how do they know where to put it to avoid things breaking?

I also suggest adding brief notes here about gotchas they might run into, e.g. avoiding Kotlin synthetic imports, direct databinding view references, fragment references within layouts, etc. People will hit these issues with new code, so we should make sure that folks know how to write things to avoid breaking Bazel or needing to dig into the problem for multiple hours.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @miaboloix! Overall, I think this looks good. I had a couple of nits, and one comment around generally improving the documentation across all BUILD files (especially app) to guard against people making mistakes as they make future changes. Let me know if you'd rather do this touch-up work in a follow-up PR, instead.

File: app/src/main/java/org/oppia/app/player/audio/LanguageInterface.kt
Comment (BenHenning) [reply app/src/main/java/org/oppia/app/player/audio/LanguageInterface.kt]: Much clearer--thanks!

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughActivityListener.kt
Comment (miaboloix) [reply app/src/main/java/org/oppia/app/walkthrough/WalkthroughActivityListener.kt]: Done

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughPageChanger.kt
Comment (miaboloix) [reply app/src/main/java/org/oppia/app/walkthrough/WalkthroughPageChanger.kt]: Okay, thank you!

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: Okay! Let me know what you think of the documentation I added!

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: Suggest striking the 'do not place' unless you're concerned with people specifically doing that in certain cases. I'm hoping that the 'place your file here' implies sufficient exclusivity that, when compared with other lists & their instructions, it should be clear to developers where to put new files when they need to.

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: Here & above: it's a bit confusing that you're referring to both lists of files in both locations. Suggest keeping the documentation specific to the thing it's documenting.

Ditto elsewhere.

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: Here & above: suggest framing the 'place your file here' instructions for view models to crisply specify when a file should go here or above, rather than having the author need to go and read about the genrule in order to understand.

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel:269]: It's a bit confusing that both this & view models can depend on resources. How do we decide where non-views/view models that depend on resources go if they could go in either list?

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: Ditto here--let's keep the 'place your file here' instructions sufficiently clear to help the author choose.

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: What about binding adapters that don't fit one of the two above requirements?

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: Should we just mention broadly that things shouldn't use databinding views or synthetic imports? Where?

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: Similar to an earlier comment, but it's also confusing to repeat the same info here. I'm hoping that the final state of this documentation will be such that: a contributor reads the documentation and knows exactly where to put a new file.

Ditto for the recycler view one below.

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: Given the requirements, maybe we should rename these to be DATABINDING_ADAPTERS_WITH_RESOURCE_IMPORTS and DATABINDING_ADAPTERS_WITHOUT_RESOURCE_IMPORTS to make things clearer. WDYT?

Ditto for the other 'altered' file lists.

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel:344]: Suggest also mentioning for adapter documentations that adapters _have_ to be defined in one of the adapter lists, and they have to be in Java.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @miaboloix! Took a full pass on the new documentation. PTAL.

File: app/BUILD.bazel
Comment (BenHenning) [reply app/BUILD.bazel]: Followed up with some comments. Thanks for starting it!

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: Okay

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: Okay

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: Okay

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: Added a comment about this at the top of the file

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: Okay

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel:269]: Wouldn't those files be kept in the app library if they could be? The only files that should be placed in a list like these are view files, listeners, annotation files, view models, and binding adapters, right?

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: okay!

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: I decided separating them was unnecessary - they are in one library now

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel]: I think changing `ALTERED_*` to `*_WITH_RESOURCE_IMPORTS` makes sense. But I am not so sure about adding `_WITHOUT_RESOURCE_IMPORTS` - wouldn't this be confusing? The files processed with the genrule are added to this list and they still have resource inputs. This also ends up being the list included in the `src` attribute.

File: app/BUILD.bazel
Comment (miaboloix) [reply app/BUILD.bazel:344]: Ok

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]:
```suggestion
Source files for the views library that import resources.
```

File: app/BUILD.bazel
Comment (BenHenning) [reply app/BUILD.bazel:269]: Okay, that makes sense and is clearer per your new comment at the top of the file & the revised per-list documentation.

File: app/BUILD.bazel
Comment (BenHenning) [reply app/BUILD.bazel]: I think what you added is clear. Thanks!

**Review by BenHenning — APPROVED** [review]
Thanks @miaboloix! Just one comment, otherwise LGTM. Feel free to resolve the new comment thread once addressed.

**Comment by anandwana001** [comment]
@miaboloix One crash we are getting after this merge. 
Firstly, in our `land/profile_chooser_profile_view` firstly it is `wasProfileEverBeenAddedValue` which should be `hasProfileEverBeenAddedValue` just like it's other files. 
Secondly, `hasProfileEverBeenAddedValue` this value is not getting updated from the presenter when we run the app which causes the crash. 

```
<variable
      name="hasProfileEverBeenAddedValue"
      type="androidx.databinding.ObservableField&lt;Boolean&gt;" />
```

**Comment by anandwana001** [comment]
Also, I am unable to find `~/bazel-databinding-new`, as per the doc after clone what I have is `~/oppia-bazel/bazel-bin/src/`.

**Pull request merged**
